storage/reservation.go

```diff
@@ -0,0 +1,247 @@
+func (pq bookingPQ) peek() *booking {
+	if len(pq) == 0 {
+		return nil
+	}
+	return (pq)[0]
+}
+
+// enqueue either adds the booking to the queue or updates its location in the
+// priority queue.
+func (pq *bookingPQ) enqueue(resTimeout *booking) {
+	if resTimeout.index < 0 {
+		heap.Push(pq, resTimeout)
+	} else {
+		heap.Fix(pq, resTimeout.index)
+	}
+}
+
+// dequeue removes the next booking from the priority queue.
+func (pq *bookingPQ) dequeue() *booking {
+	if len(*pq) == 0 {
+		return nil
+	}
+	return heap.Pop(pq).(*booking)
+}
+
+// bookie contains a store's replica reservations.
+type bookie struct {
+	clock              *hlc.Clock
+	reservationTimeout time.Duration // How long each reservation is held.
+	mu                 struct {
+		sync.RWMutex                              // Protects all values within the mu struct.
+		queue        bookingPQ                    // Priority Queue used to handle expiring of reservations.
+		resByRangeID map[roachpb.RangeID]*booking // All active reservations
+		size         int64                        // Total bytes required for all reservations.
+	}
+}
+
+// newBookie creates a reservations system and starts its timeout queue.
+func newBookie(clock *hlc.Clock, reservationTimeout time.Duration, stopper *stop.Stopper) *bookie {
+	b := &bookie{
+		clock:              clock,
+		reservationTimeout: reservationTimeout,
+	}
+	b.mu.resByRangeID = make(map[roachpb.RangeID]*booking)
+	heap.Init(&b.mu.queue)
+	b.start(stopper)
+	return b
+}
+
+// len returns the total number of outstanding reservations.
+func (b *bookie) len() int {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	return len(b.mu.resByRangeID)
+}
+
+// size returns the total bytes reserved.
+func (b *bookie) size() int64 {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	return b.mu.size
+}
+
+// reserve resreves a new replica.
+func (b *bookie) reserve(res reservation) bool {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if oldRes, ok := b.mu.resByRangeID[res.rangeID]; ok {
+		if log.V(2) {
+			log.Infof("there is already an existing reservation for rangeID:%d, %v", res.rangeID,
+				oldRes)
+		}
+		return false
+	}
+
+	newBooking := &booking{
+		reservation: res,
+		expireAt:    b.clock.Now().Add(b.reservationTimeout.Nanoseconds(), 0),
+		index:       -1, // This is required to add it to the queue.
+	}
+
+	b.mu.resByRangeID[res.rangeID] = newBooking
+	b.mu.queue.enqueue(newBooking)
+	b.mu.size += res.size
+	return true
+}
+
+// fill removes a reservation.
+func (b *bookie) fill(rangeID roachpb.RangeID) bool {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	// Lookup the reservation.
+	res, ok := b.mu.resByRangeID[rangeID]
+	if !ok {
+		if log.V(2) {
+			log.Infof("there is no reservation for rangeID:%d", rangeID)
+		}
+		return false
+	}
+
+	return b.fillBookingLocked(res)
+}
+
+// fillBookingLocked fills a booking. It requires that the reservation lock is
+// held. This should only be called internally.
```

Comment:
consider exporting the methods which you expect to be called externally (they won't show up in godoc since the receiver isn't exported).